### PR TITLE
TOTP implementation

### PR DESCRIPTION
**Pull request by chris-4chain**
Hello :) 
According to our conversation with @sirdeggen I'm putting here the entire `TOTP for PIKE` algorithm from `spv-wallet-js-client`. It's a draft which might be a good starting point to discuss what to move to `ts-sdk` and what to keep in `spv-wallet-js-client` (regarding TOTP stuff).

## Context
We faced some issues during implementation of TOTP codes for javascript client. First attempt was to just use some OTP library, like [otplib](https://github.com/yeojz/otplib) - and it worked fine, but only for node.js (and web - depends on what lib we were trying). 
However, it didn't work on react-native mobile app. There were issues with `Buffer` and usage of `crypto/crypto-js`.  Of course, polyfills would solve these issues - but we didn't like this idea. 

## Implementation
Due to these challenges, and since `ts-sdk` already had the necessary core algorithms (SHA1/SHA256/SHA512) implemented, we decided to implement our own TOTP algorithm.

One thing was missing: `HMAC` for `SHA1` algorithm, which is a default one for TOTP codes - so we implemented it, based on the code for `SHA256HMAC` and with usage of already implemented `SHA1`. 

You can check the implementation here: https://github.com/bitcoin-sv/spv-wallet-js-client/tree/main/src/totp 

## What's inside
- missing `SHA1HMAC` algorithm (which would be useful to have here - not in `spv-wallet-js-client`)
- `totp.ts` - with the implementation of standard TOTP codes (because of it provides a functionality, similar to what `otplib` offers - it might be beneficial to move it to `ts-sdk`)
- `pike-totp.ts` - A specialized algorithm for generating TOTP codes based on a client’s `xPriv` and PKI of some contact. It's built on top of standard TOTP generator/validator (I'm not sure if it should be moved to `ts-sdk` - What do you think 🤔?)
- several utility functions - which maybe have its equivalents in ts-sdk but we didn't found them 

## Adjustments
Just for the record. The code from `spv-wallet-js-client` slightly differs from what is in this PR, because I've made some adjustments to make it work within this project.

### CODE REVIEW

**Comment by sirdeggen** [comment]
@ty-everett keen for your input.

I think TOTP is a useful method to add to the SDK as it solves a core problem - validation of remote counterparties when using hosted services which are subject to man in the middle attacks.

I recommend we add `totp.ts` but leave the PIKE specific implementation to the SPV Wallet codebase since it uses xPrivs which I'd rather change there than introduce here.

I'll review and contribute any suggestions I have in the coming days.

File: src/totp/converters.ts
Comment (sirdeggen) [inline src/totp/converters.ts:1]: We almost certainly have this elsewhere `toArray()` comes to mind.

File: src/totp/totp.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+};
+
+export class TOTP {
+  /**
+   * Generates a Time-based One-Time Password (TOTP).
+   * @param {Uint8Array} rawKey - The secret key for TOTP.
+   * @param {TOTPOptions} options - Optional parameters for TOTP.
+   * @returns {string} The generated TOTP.
+   */
+  static generate(rawKey: Uint8Array, options?: TOTPOptions): string {
```
Comment (sirdeggen) [inline src/totp/totp.ts:35]: raw prefix we could probably drop, and I'd name it `secret` instead to match the language of the rest of the sdk.

File: src/totp/types.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export interface Contact {
+  /**
+   * Paymail related to contact
+   */
+  paymail: string;
```
Comment (sirdeggen) [inline src/totp/types.ts:5]: I don't think we should build anything specific to paymail into this lib, there's a ts-paymail for that. Looking for how this is used.

Comment (chris-4chain) [reply src/totp/types.ts:5]: Yes, it was necessary for `pike-totp` algorithm to make it work in here. 
The actual type in js-client is like this: 
https://github.com/bitcoin-sv/spv-wallet-js-client/blob/main/src/types.ts#L154-L191

**Review by sirdeggen — APPROVED** [review]
I made some changes and would approve of this being added to the SDK.

**Comment by ty-everett** [comment]
I feel this may be beyond the scope of this project (scope creep), as it feels implementation specific. If we support it here, we might misalign with other ways of doing similar things, and we're also implicitly committing to maintain this across at least Python and Go in the future.

**Review by sirdeggen — APPROVED** [review]

**Pull request merged**